JSONObject.java

```diff
@@ -578,10 +581,10 @@ public double getDouble(String key) throws JSONException {
         Object object = this.get(key);
         try {
             return object instanceof Number ? ((Number) object).doubleValue()
-                    : Double.parseDouble((String) object);
+                    : new BigDecimal((String) object).doubleValue();
```

Comment:
No, `Double.parse` will throw an exception if the number can not be represented as a double.
From the javadocs:

> double java.lang.Double.parseDouble(String s) throws NumberFormatException
> 
> Returns a new double initialized to the value represented by the specified String, as performed by the valueOf method of class Double.
> Parameters:s the string to be parsed.Returns:the double value represented by the string argument.Throws:NullPointerException - if the string is nullNumberFormatException - if the string does not contain a parsable double.Since:1.2See Also:java.lang.Double.valueOf(String)
> 
> 

`Double.valueOf` seems to do something closer to what you are trying to accomplish, but it would be the same amount of garbage collection.

`new BigDecimal(String)` will always parse the value as long as it's a valid decimal number (i.e. not "abc"). The `doubleValue` call will then do a narrowing conversion which may change the value to -Inf or +Inf.

Lastly, this change would be out of scope for this PR. the `get*` methods were not modified here, only the opt*. If you feel we should evaluate it further, feel free to open another pull request with the changes and notes to performance impacts.